src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateClones = [Object.assign({}, state)];
+
+  for (const action of actions) {
+    let newState = Object.assign({}, stateClones[stateClones.length - 1]);
+
+    switch (action.type) {
+      case 'addProperties':
+        newState = Object.assign(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const keyToRemove of action.keysToRemove) {
+          delete newState[keyToRemove];
+        }
+        break;
+
+      case 'clear':
+        newState = {};
+        break;
+    }
+
+    stateClones.push(newState);
+  }
+  stateClones.shift();
```

Comment:
You can use `...spread` instead of `Object.assign` for easier reading and not use `shift()` on every iteration